File: ssh-copy-id.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char *argv[])
{
    if (argc < 3)
    {
        printf("Usage: %s user@machine [ssh arguments]\n", argv[0]);
        return 1;
    }

    char publicKey[256];
    snprintf(publicKey, sizeof(publicKey), "%s/.ssh/id_rsa.pub", getenv("USERPROFILE"));

    FILE *publicKeyFile = fopen(publicKey, "r");
    if (!publicKeyFile)
    {
        printf("ERROR: failed to open ID file '%s': No such file\n", publicKey);
        return 1;
    }

    char sshCommand[512];
    snprintf(sshCommand, sizeof(sshCommand), "ssh %s \"umask 077; test -d .ssh || mkdir .ssh ; cat >> .ssh/authorized_keys || exit 1\"", argv[2]);

    FILE *sshProcess = popen(sshCommand, "w");
    if (!sshProcess)
    {
        printf("Failed to start ssh process\n");
        fclose(publicKeyFile);
        return 1;
    }

    char buffer[1024];
    size_t bytesRead;

    while ((bytesRead = fread(buffer, 1, sizeof(buffer), publicKeyFile)) > 0)
    {
        fwrite(buffer, 1, bytesRead, sshProcess);
    }

    fclose(publicKeyFile);
    pclose(sshProcess);

    printf("Successfully copied public key to %s\n", argv[1]);

    return 0;
}

File: ssh-copy-id.cs
using System;
using System.Diagnostics;
using System.IO;

class Program
{
    static void Main(string[] args)
    {
        string userAtMachine = args[0];
        string sshArgs = string.Join(" ", args, 1, args.Length - 1);

        string publicKey = Environment.GetEnvironmentVariable("USERPROFILE") + "/.ssh/id_rsa.pub";
        if (!File.Exists(publicKey))
        {
            Console.WriteLine("ERROR: failed to open ID file '{0}': No such file", publicKey);
            return;
        }

        ProcessStartInfo startInfo = new ProcessStartInfo
        {
            FileName = "ssh",
            Arguments = string.Format("{0} \"umask 077; test -d .ssh || mkdir .ssh ; cat >> .ssh/authorized_keys || exit 1\"", sshArgs),
            RedirectStandardInput = true
        };

        using (Process process = Process.Start(startInfo))
        {
            using (StreamWriter writer = process.StandardInput)
            {
                using (StreamReader reader = new StreamReader(publicKey))
                {
                    writer.Write(reader.ReadToEnd());
                }
            }

            process.WaitForExit();

            if (process.ExitCode == 0)
            {
                Console.WriteLine("Successfully copied public key to {0}", userAtMachine);
            }
            else
            {
                Console.WriteLine("Failed to copy public key");
            }
        }
    }
}
